### PR TITLE
TreeSelect: remove margin overrides and add new opt-in prop

### PR DESCRIPTION
**Pull request by brookewp**
## What?

Added new opt-in prop `__nextHasNoMarginBottom` for usages of `TreeControl` in the Gutenberg codebase and removed margin overrides. 

## Why?

Part of this project: https://github.com/WordPress/gutenberg/issues/38730
The tl;dr is `BaseControl` has a `margin-bottom` which makes it difficult to reuse and results in inconsistent use. 

## How?

By removing margin overrides in the CSS and adding the prop `__nextHasNoMarginBottom`.

### HierarchicalTermSelector

1. Edit a post
2. Look for 'Categories' in the block inspector
3. Click on 'Add New Category'
4. Verify the spacing 

| Before  | After |
| ------------- | ------------- |
|  Screenshot 2023-01-27 at 3 35 31 PM | Screenshot 2023-01-27 at 3 32 00 PM |

### CategorySelect & AuthorSelect

I don't believe that `CategorySelect` shows up anywhere/matches any conditions based on [this part of the readme](https://github.com/WordPress/gutenberg/blob/trunk/packages/components/src/query-controls/README.md#multiple-category-selector):

>The `QueryControls` component now supports multiple category selection, to replace the single category selection... To enable it use the component with the new props instead: `categorySuggestions` in place of `categoriesList` 

So to test it, the only way I found was to replace `categorySuggestions={ categorySuggestions }` with `categoriesList={ categories }` on line 343 in `packages/block-library/src/latest-posts/edit.js`: https://github.com/WordPress/gutenberg/blob/6a64cdfee23442ee201ea5039ae5c833d1946428/packages/block-library/src/latest-posts/edit.js#L343

Then:

1. Add the Latest Posts block 
2. Click on the settings icon in the block inspector
3. Look for 'Category' and 'Author' within the 'Sorting and filtering' panel
4. Ensure the space below those select elements is the same as before 

Screenshot 2023-01-27 at 2 46 48 PM

### CODE REVIEW

File: packages/editor/src/components/post-taxonomies/hierarchical-term-selector.js
Comment (brookewp) [inline packages/editor/src/components/post-taxonomies/hierarchical-term-selector.js:12]: When adding the prop to `TreeSelect`, a visible amount of space is lost below the 'Add new category' submit button.


| Before prop  | After prop |
| ------------- | ------------- |
|  Screenshot 2023-01-27 at 3 44 26 PM | Screenshot 2023-01-27 at 3 44 02 PM  |

We could separate [the CSS selectors for the add and submit buttons](https://github.com/WordPress/gutenberg/blob/6a64cdfee23442ee201ea5039ae5c833d1946428/packages/editor/src/components/post-taxonomies/style.scss#L21-L24) or remove the CSS and use components instead. In this case, it worked better to use `Flex` over `VStack` so we could use `FlexItem` to ensure that the buttons retain their width, both for their size and their focus outlines. 

If the goal is to use VStack/HStack over Flex/FlexItem/FlexBlock in the future, I wonder if it would be worth adding a way to self-align the children? 🤔

File: packages/editor/src/components/post-taxonomies/style.scss
Comment (brookewp) [inline packages/editor/src/components/post-taxonomies/style.scss:20]: I saw both pixel values and variables used, so I updated everything to variables with the corresponding value while I was already here. 🙂

File: packages/editor/src/components/post-taxonomies/hierarchical-term-selector.js
Comment (mirka) [reply packages/editor/src/components/post-taxonomies/hierarchical-term-selector.js:12]: What do you think about just using `FlexItem` inside VStack/HStack? Too weird?

File: packages/components/src/query-controls/author-select.tsx
Comment (mirka) [inline packages/components/src/query-controls/author-select.tsx:30]: Looks like this won't quite work outside of the block inspector. Could you check in the Storybook to see if it actually needs a VStack?

Uneven spacing in the Storybook instance for QueryControls

File: packages/editor/src/components/post-taxonomies/hierarchical-term-selector.js
Comment (ciampo) [inline packages/editor/src/components/post-taxonomies/hierarchical-term-selector.js:429]: Should we prefer a `<div />` (or any other HTML element) with custom CSS flex styles here, as mentioned in https://github.com/WordPress/gutenberg/issues/43085 ? Or should we continue using `Flex` for now, and look into migrating away from it at a later point? (cc @mirka )

File: packages/editor/src/components/post-taxonomies/hierarchical-term-selector.js
Comment (brookewp) [reply packages/editor/src/components/post-taxonomies/hierarchical-term-selector.js:12]: Is it okay to do that? I assumed it wouldn't be since the docs mention using FlexItem/Block inside Flex. But if it's not a faux pas then I could do that. 😄

File: packages/editor/src/components/post-taxonomies/hierarchical-term-selector.js
Comment (ciampo) [reply packages/editor/src/components/post-taxonomies/hierarchical-term-selector.js:429]: Seeing that we may use `FlexItem` as suggested in https://github.com/WordPress/gutenberg/pull/47515#discussion_r1093269132, we can probably also keep using `Flex` for the time being.

We can have a look at "dismissing" `Flex` and its subcomponents at a later point

Comment (mirka) [reply packages/editor/src/components/post-taxonomies/hierarchical-term-selector.js:429]: I guess the pro/cons to using `Flex` are similar to Tailwind, where you get dramatically less CSS data in the final output in return for having to learn a new set of styling conventions. I don't have too strong of an opinion one way or the other, but I slightly lean towards using `Flex` for now because it will probably be easier to refactor once we are ready to move to a better spacing component (#43423).

Comment (ciampo) [reply packages/editor/src/components/post-taxonomies/hierarchical-term-selector.js:429]: Sounds good!

File: packages/editor/src/components/post-taxonomies/hierarchical-term-selector.js
Comment (mirka) [reply packages/editor/src/components/post-taxonomies/hierarchical-term-selector.js:12]: I mean, it technically works but it is definitely debatable whether it should be a faux pas! VStack/HStack are just `Flex` components under the hood with some preset props for convenience. The perceived ickiness of using something called a "FlexItem" inside a "VStack" is what we often call a _leaky abstraction_. The Stack components are intended to abstract (i.e. hide) the notion of a flexbox into something less technical, but that abstraction leaks when one is forced to confront with flex technicalities.

So yeah, I am not pressing the matter, it's a gray area 🙃

File: packages/components/src/query-controls/author-select.tsx
Comment (brookewp) [reply packages/components/src/query-controls/author-select.tsx:30]: If we add VStack, then a couple of things come up:

1. In Storybook, all looks good except for the help text. Is it worth adding `StyledHelp` to `FormTokenField`? It already has `StyledLabel`. I would do that in a separate PR if this makes sense. 
2. In the editor, while looking at the Latest Posts block, the spacing is too large with VStack (due to the inspector's 24px margin bottom styles). It would need `spacing="0",` but then QueryControls in Storybook would end up looking the same as your screenshot above. We could add a CSS override to the Latest Posts block but then I believe we'd need to add a CSS class to QueryControls? 

@mirka thoughts? 

**QueryControls in Storybook:**

| Trunk  | With VStack |
| ------------- | ------------- |
| Screenshot 2023-02-06 at 7 18 41 PM  | Screenshot 2023-02-06 at 7 21 50 PM |

**QueryControls in Latest Posts block:**

| Trunk  | With VStack |
| ------------- | ------------- |
| Screenshot 2023-02-06 at 7 14 59 PM | Screenshot 2023-02-06 at 7 14 49 PM |

File: packages/editor/src/components/post-taxonomies/hierarchical-term-selector.js
Comment (brookewp) [reply packages/editor/src/components/post-taxonomies/hierarchical-term-selector.js:12]: >we often call a leaky abstraction.

Always love to learn new terms. :) 

From [your other comment](https://github.com/WordPress/gutenberg/pull/47515#discussion_r1095795419):
>I slightly lean towards using Flex for now because it will probably be easier to refactor once we are ready to move to a better spacing component 

To confirm, are you now saying to leave this as Flex to make it easier to move later?

File: packages/components/src/query-controls/author-select.tsx
Comment (mirka) [reply packages/components/src/query-controls/author-select.tsx:30]: > 1. In Storybook, all looks good except for the help text. Is it worth adding `StyledHelp` to `FormTokenField`? It already has `StyledLabel`. I would do that in a separate PR if this makes sense.

Sure! FormTokenField actually needs to have its margin deprecated (part of #39358). Migrating to `StyledHelp` would be part of that.

> 2. In the editor, while looking at the Latest Posts block, the spacing is too large with VStack (due to the inspector's 24px margin bottom styles). It would need `spacing="0",` but then QueryControls in Storybook would end up looking the same as your screenshot above. We could add a CSS override to the Latest Posts block but then I believe we'd need to add a CSS class to QueryControls?

QueryControls itself should only care about looking correct in isolation (Storybook), not in the block editor. The block editor needs to counteract any of its own margin hacks on its own. That would likely be in [this file](https://github.com/WordPress/gutenberg/blob/805618c9ee1f45eb0634a723ad903b87fb19d992/packages/block-editor/src/components/block-inspector/style.scss#L23), like we recently did for FocalPointPicker. I _think_ we wouldn't need to add new CSS classes to FormTokenField itself, but let me know if that's not the case.

File: packages/editor/src/components/post-taxonomies/hierarchical-term-selector.js
Comment (mirka) [reply packages/editor/src/components/post-taxonomies/hierarchical-term-selector.js:12]: In that comment, I was only advocating for using `<Flex>` over a "`<div>` with handwritten flexbox CSS". 

Whether to use `Flex`/`VStack` in this case with FlexItem, I don't have a strong opinion. I'll leave it up to your preferences 😄

File: packages/components/src/query-controls/author-select.tsx
Comment (brookewp) [reply packages/components/src/query-controls/author-select.tsx:30]: >QueryControls itself should only care about looking correct in isolation (Storybook), not in the block editor. 

Thank you! This is a good reminder. ❤️ 

> I think we wouldn't need to add new CSS classes to FormTokenField itself, but let me know if that's not the case.
Sorry, I meant the Latest Posts block. 

I've updated `QueryControls` so it looks better in Storybook. The spacing should still be adjusted for the help below `FormTokenField` but I don't think it's too bad. And it currently looks better than it does in trunk, due to the lack of even spacing above categories. So I figured it may be better to get this out sooner than later.

File: packages/editor/src/components/post-taxonomies/hierarchical-term-selector.js
Comment (brookewp) [reply packages/editor/src/components/post-taxonomies/hierarchical-term-selector.js:12]: Thanks for clarifying! I'll leave it as Flex, as a 'leaky abstraction' doesn't sound ideal. Also, I wouldn't want to influence others to start doing that if they came across this instance. 😅

File: packages/block-editor/src/components/block-inspector/style.scss
Comment (mirka) [inline packages/block-editor/src/components/block-inspector/style.scss]: To clarify, what we want is to "un-apply" or in other words "opt out" the QueryControl components from these overrides:

https://github.com/WordPress/gutenberg/blob/12e2cdeea5b36ea9b542fc71662a46661029027a/packages/block-editor/src/components/block-inspector/style.scss#L13-L19

That means to apply margin-bottom `0` (or `auto`), which is the neutral state. Similar to what we're doing for the FocalPickerControl just above. This "blank slate" allows us to use clean flex gap styling within QueryControl itself. The CSS scope is `.components-base-control` in all `QueryControl`s, not just in the Latest Posts block.

I may have misled you when I said we probably didn't have to add new CSS classes. I was misreading [these `key`s ](https://github.com/WordPress/gutenberg/blob/12e2cdeea5b36ea9b542fc71662a46661029027a/packages/components/src/query-controls/index.tsx#L84) as class names 😆

This may be conceptually confusing, so feel free to ping me if you'd like to discuss further!

Comment (brookewp) [reply packages/block-editor/src/components/block-inspector/style.scss]: >That means to apply margin-bottom 0 (or auto), which is the neutral state. Similar to what we're doing for the FocalPickerControl just above. This "blank slate" allows us to use clean flex gap styling within QueryControl itself.

Thanks for clarifying, @mirka! This is clear to me. 

>The CSS scope is .components-base-control in all QueryControls, not just in the Latest Posts block.

And this makes sense, _but_ I have a few questions. 😄

 In this PR, `QueryControls` appears to match trunk in Storybook when adding `VStack` with a gap of 8px. But if we remove the `BasecControl` inspector override as we did with `FocalPickerControl`, the spacing looks off in the Latest Posts block (screenshot below), due to other panels in the inspector that still have those bottom margins. So would we then need to add a specific CSS override to the Latest Posts block after to fix the issue this would cause? 

Screenshot 2023-02-14 at 5 32 41 PM

I believe the Latest Posts block is the only usage of `QueryControls` in GB. Isn't the `BaseControl` override only an issue within the inspector within GB? Or are there other areas that could be affected?

File: packages/editor/src/components/post-taxonomies/style.scss
Comment (ciampo) [inline packages/editor/src/components/post-taxonomies/style.scss]: I couldn't find any element with class `.editor-post-taxonomies__hierarchical-terms-label` in the code — maybe these styles are not used anymore and can be deleted?

File: packages/block-editor/src/components/block-inspector/style.scss
Comment (ciampo) [reply packages/block-editor/src/components/block-inspector/style.scss]: You make a good point — it seems like we have two choices:

1. Go ahead with the solution suggested by @mirka — good in principle, but it may introduce some visual inconsistencies when looking at the Latest Posts controls
2. Leave the `BaseControl` overrides as they are currently.

We could opt for 2, and remove the `BaseControl` margins separately in the future?

Finally, I just noticed that `QueryControls` doesn't seem to have a root classname that we could use anyway (i.e. `components-query-control`).

Comment (mirka) [reply packages/block-editor/src/components/block-inspector/style.scss]: You're asking whether it's ok to have 8px margins for QueryControl controls while the rest of the Latest Posts controls have 24px margins, correct?

Our [new controls grid](https://github.com/WordPress/gutenberg/issues/43423#issuecomment-1232540038) is 16px gap (not 8px), so just using 16px for the VStack already will ease the awkwardness a bit.

I don't have a strong opinion on what the exact margin values should be in the Latest Posts block inspector, but I think the minimal requirements we should hit in this PR are:

- It looks correct in the Storybook (16px gap).
- Any QueryControl component injected into the block inspector sidebar should look correct (not just in Latest Posts, because e.g. third-party blocks could use it).

What do you think?

Comment (brookewp) [reply packages/block-editor/src/components/block-inspector/style.scss]: _ **tl;dr**, if we have it look correct in storybook, it won't look good for the Latest Posts block, which appears to be the only use case of `QueryControls` in GB._ 

Aside from what I've done in this PR and #48609 to fix the spacing below the help text, I'm not sure what the next step is. 😕 

>You're asking whether it's ok to have 8px margins for QueryControl controls while the rest of the Latest Posts controls have 24px margins, correct?

@mirka, I'm not sure if this was directed at me or at @ciampo, but I'll share my thoughts in case I wasn't clear.

I had set the spacing to 8px, to match what `QueryControls` looks like in trunk. If it makes sense to update to the new controls grid spacing, then I'm happy to do that (well I already did in my most recent commit 😂 )

My original concern was the result on the Latest Posts block, with either the 8px or 16px spacing because with the override, the space becomes massive:

Screenshot 2023-02-28 at 7 29 38 PM

>It looks correct in the Storybook (16px gap).
>Any QueryControl component injected into the block inspector sidebar should look correct (not just in Latest Posts, because e.g. third-party blocks could use it).

Your minimum requirements make perfect sense. In regards to third-parties, would the block inspector override affect them or is it contained to GB? 

I thought it was the latter, which is why I thought we could add the override in this case/in the meantime, while we can't remove the margin from the inspector.

Comment (mirka) [reply packages/block-editor/src/components/block-inspector/style.scss]: What I had in mind is pretty simple — to adopt the same "opt out" pattern we established in FocalPointPicker. Something like this 👉 https://github.com/WordPress/gutenberg/commit/6cd387ebbe9dcf153656f80fdcf4b92eec9fd25d

> In regards to third-parties, would the block inspector override affect them or is it contained to GB?

The former. The block inspector override will affect third-party usages of QueryControls as long as they are rendered in the block inspector. This is why we can't just patch it for Latest Posts, it has to work for all QueryControls.

Comment (brookewp) [reply packages/block-editor/src/components/block-inspector/style.scss]: Thanks for the context and the commit! When reviewing the commit, I see the option to pull the request. Is that the next step here? (also need to rebase due to changelog). I haven't seen this before, so just wanted to check!

File: packages/editor/src/components/post-taxonomies/style.scss
Comment (brookewp) [reply packages/editor/src/components/post-taxonomies/style.scss]: It looks like it! I can update this

File: packages/block-editor/src/components/block-inspector/style.scss
Comment (mirka) [reply packages/block-editor/src/components/block-inspector/style.scss]: Good question! You can do whatever you want with it, including pulling or cherry-picking or just taking inspiration from the diff 😄 Another fun tip is to append [`.patch` to the commit URL](https://github.com/WordPress/gutenberg/commit/6cd387ebbe9dcf153656f80fdcf4b92eec9fd25d.patch) and you'll get a [patch file](https://gist.github.com/nepsilon/22bc62a23f785716705c).

Comment (ciampo) [reply packages/block-editor/src/components/block-inspector/style.scss]: > Another fun tip is to append [.patch to the commit URL](https://github.com/WordPress/gutenberg/commit/6cd387ebbe9dcf153656f80fdcf4b92eec9fd25d.patch) and you'll get a [patch file](https://gist.github.com/nepsilon/22bc62a23f785716705c).

🤯

File: packages/block-library/src/latest-posts/edit.js
Comment (mirka) [inline packages/block-library/src/latest-posts/edit.js]: Classname not needed anymore? 🧹

**Review by mirka — APPROVED** [review]
Looks good! Ready to merge once the unneeded CSS class is removed 🚀

**Pull request merged**
